### PR TITLE
Expose the complete API for dealing with KV cache and states

### PR DESCRIPTION
**Pull request by zh217**
This pull request exposes the complete public C API for dealing with KV cache and internal states in Rust. I need this for writing efficient custom decoder, and it might also be useful for others.

The added method `copy_kv_cache_seq` is almost the existing `copy_cache`, but the latter is missing one argument. So I have added a note in the documentation of the latter.

### CODE REVIEW

File: llama-cpp-2/src/context/kv_cache.rs
Comment (MarcusDunn) [inline llama-cpp-2/src/context/kv_cache.rs:128]: What does a negative seq_id mean here (and elsewhere) Is it ever valid / meaningful?

File: llama-cpp-2/src/context/kv_cache.rs
Comment (MarcusDunn) [inline llama-cpp-2/src/context/kv_cache.rs:118]: Avoid `as`, prefer `i32:from` as it'll throw an error if `d`s type ever changes to something that could fail.

File: llama-cpp-2/src/context/session.rs
Comment (MarcusDunn) [inline llama-cpp-2/src/context/session.rs:157]: make `src` mutable. I'll open a PR to llama.cpp to make it `const` on their end

Comment (MarcusDunn) [reply llama-cpp-2/src/context/session.rs:157]: you can document that `src` will be changed to be immutable pending upstream changes, I'll fix once PR is merged.

**Review by MarcusDunn — COMMENTED** [review]
these are all great to add.

- remove or deprecate the old API
- document what negative numbers mean or change them to be unsigned (or if they are valid leave as is)
- nit: remove all `as` in favor of `TryFrom` or `From` (panicking is fine for now)

File: llama-cpp-2/src/context/session.rs
Comment (MarcusDunn) [reply llama-cpp-2/src/context/session.rs:157]: or wait until PR is merged and the nightly update llama.cpp goes though.

File: llama-cpp-2/src/context/kv_cache.rs
Comment (zh217) [reply llama-cpp-2/src/context/kv_cache.rs:128]: As far as I understand, `seq_id` is just a label, ultimately used in a `std::set`, so yes negative numbers are fine.

In the C++ code almost all such parameters actually refers to `typedef int32_t llama_seq_id`, however the existing Rust wrapper does not introduce a newtype (notably `LlamaBatch::add`).

Also, I don't think there is enough checking in the C++ code to check what is passed in makes sense (a proper check would be highly non-trivial and slow). Worse, passing in the wrong thing is not guaranteed to trigger anything noticable. So I'm not sure what is the best way to do in the Rust part.

**Review by MarcusDunn — APPROVED** [review]
Looks good. Pending on llama.cpp changes before I cut a release but thanks for all the work!

**Pull request merged**
